// stryker-typescript/src/TypescriptMutator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var flatMap = require("lodash.flatmap");
var tsHelpers_1 = require("./helpers/tsHelpers");
var plugin_1 = require("stryker-api/plugin");
var mutator_1 = require("./mutator");
function typescriptMutatorFactory(injector) {
    return injector
        .provideValue(exports.MUTATORS_TOKEN, mutator_1.nodeMutators)
        .injectClass(TypescriptMutator);
}
exports.typescriptMutatorFactory = typescriptMutatorFactory;
typescriptMutatorFactory.inject = plugin_1.tokens(plugin_1.commonTokens.injector);
exports.MUTATORS_TOKEN = 'mutators';
var TypescriptMutator = /** @class */ (function () {
    function TypescriptMutator(options, mutators) {
        this.options = options;
        this.mutators = mutators;
    }
    TypescriptMutator.prototype.mutate = function (inputFiles) {
        var _this = this;
        var tsConfig = tsHelpers_1.getTSConfig(this.options);
        var mutants = flatMap(inputFiles, function (inputFile) {
            var sourceFile = tsHelpers_1.parseFile(inputFile, tsConfig && tsConfig.options && tsConfig.options.target);
            return _this.mutateForNode(sourceFile, sourceFile);
        });
        return mutants;
    };
    TypescriptMutator.prototype.mutateForNode = function (node, sourceFile) {
        var _this = this;
        var targetMutators = this.mutators.filter(function (mutator) { return mutator.guard(node); });
        var mutants = flatMap(targetMutators, function (mutator) { return mutator.mutate(node, sourceFile); });
        node.forEachChild(function (child) {
            // It is important that forEachChild does not return a true, otherwise node visiting is halted!
            mutants.push.apply(mutants, _this.mutateForNode(child, sourceFile));
        });
        return mutants;
    };
    TypescriptMutator.inject = plugin_1.tokens(plugin_1.commonTokens.options, exports.MUTATORS_TOKEN);
    return TypescriptMutator;
}());
exports.TypescriptMutator = TypescriptMutator;
